pandas/core/strings.py

```diff
@@ -107,50 +112,117 @@ def cat_safe(list_of_columns: List, sep: str):
+    result_dtype : {"int", "bool", "string", "object"}
+        The output dtype. For most methods (e.g. str.upper, str.isint)
+        this can be known statically. For some, (e.g. str.get) we need
+        to infer the result dtype based on the values. See `method`.
+    na_result : Any
+        The result value to use for missing values.
+    method : str, optional.
+        Most methods should *not* pass this; it's unnecessary. It's
+        only necessary for `.str` methods that don't really operate
+        (just) on strings. For this reason, it's only relevant for
+        object-dtype input, and is completely unnecessary for
+        StringArray.
+    """
+    assert result_dtype in {"int", "bool", "string", "object"}
+    arr = extract_array(arr, extract_numpy=True)
+    mask = isna(arr)
+
+    if is_extension_array_dtype(arr.dtype):
+        # This is our nice, future world.  We *always* know the output
+        # dtype based just on `result_dtype`, no need to look at
+        # the values.
+        notna = ~mask
+        anyna = np.any(mask)
+
+        if result_dtype == "int":
+            result_dtype = pandas_dtype("Int64")
+            # TODO: Avoid this object allocation
+            #   Should be able to just use zero, and pass through
+            #   the mask to IntegerArray.
+            nd_dtype = "object"
+        elif result_dtype == "bool":
+            # TODO: return a BoolArray here.
+            result_dtype = object
+            nd_dtype = object
+        else:
+            nd_dtype = object
 
-    if isinstance(arr, ABCSeries):
-        arr = arr.values
-    if not isinstance(arr, np.ndarray):
-        arr = np.asarray(arr, dtype=object)
-    if na_mask:
-        mask = isna(arr)
-        convert = not np.all(mask)
-        try:
-            result = lib.map_infer_mask(arr, f, mask.view(np.uint8), convert)
-        except (TypeError, AttributeError) as e:
-            # Reraise the exception if callable `f` got wrong number of args.
-            # The user may want to be warned by this, instead of getting NaN
-            p_err = (
-                r"((takes)|(missing)) (?(2)from \d+ to )?\d+ "
-                r"(?(3)required )positional arguments?"
+    else:
+        # This is the bad side of things for legacy support. Deprecating
+        # the .str accessor on object dtype will make much of this go away.
+        # 1. The result dtype depends on whether there are any NA values (int, bool)
+        # 2. Non-string `method`s need special handling for "invalid values" detection.
+        #    We were *very* flexible in the past about non-string data.
+        if method == "decode":
+            non_strings = ~np.asarray([isinstance(x, bytes) for x in arr], dtype=bool)
+        elif method in {"get", "join"}:
+            # we want anything that supports __getitem__ or .get
+            non_strings = ~np.asarray(
+                [isinstance(x, (Iterable, Sequence, Mapping)) for x in arr], dtype=bool
```

Comment:
Do we need to do this check? (just trying to think if users could have had objects that would work with `get` but would not be following this check)